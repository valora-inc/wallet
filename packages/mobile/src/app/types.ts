export enum SuperchargeButtonType {
  PillRewards = 'PILL_REWARDS',
  PillSupercharge = 'PILL_SUPERCHARGE',
  MenuRewards = 'MENU_REWARDS',
  MenuSupercharge = 'MENU_SUPERCHARGE',
}

export interface Dapp {
  id: string
  categoryId: string
  iconUrl: string
  name: string
  description: string
  dappUrl: string
  isFeatured: boolean
}
