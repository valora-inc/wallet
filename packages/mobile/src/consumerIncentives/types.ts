export enum SuperchargeToken {
  cUSD = 'cUSD',
  cEUR = 'cEUR',
  cREAL = 'cREAL',
}

export interface SuperchargeTokenConfig {
  token: SuperchargeToken
  minBalance: number
  maxBalance: number
}

export interface SuperchargePendingReward {
  amount: string
  contractAddress: string
  createdAt: number
  index: number
  proof: string[]
  tokenAddress: string
}
