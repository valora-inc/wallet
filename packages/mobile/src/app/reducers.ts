import { Platform } from 'react-native'
import { Actions, ActionTypes, AppState } from 'src/app/actions'
import { SuperchargeButtonType } from 'src/app/types'
import { REMOTE_CONFIG_VALUES_DEFAULTS } from 'src/firebase/remoteConfigValuesDefaults'
import { Screens } from 'src/navigator/Screens'
import { getRehydratePayload, REHYDRATE, RehydrateAction } from 'src/redux/persist-helper'

export interface State {
  loggedIn: boolean
  numberVerified: boolean
  analyticsEnabled: boolean
  requirePinOnAppOpen: boolean
  appState: AppState
  locked: boolean
  lastTimeBackgrounded: number
  sessionId: string
  minVersion: string | null
  celoEducationUri: string | null
  celoEuroEnabled: boolean
  dappListApiUrl: string | null
  inviteModalVisible: boolean
  activeScreen: Screens
  hideVerification: boolean
  showRaiseDailyLimitTarget: string | undefined
  walletConnectV1Enabled: boolean
  walletConnectV2Enabled: boolean
  rewardsPercent: number
  rewardsStartDate: number
  rewardsMax: number
  rewardsMin: number
  rewardsABTestThreshold: string
  // In 1.13 we had a critical error which requires a migration to fix. See |verificationMigration.ts|
  // for the migration code. We can remove all the code associated with this after some time has passed.
  ranVerificationMigrationAt: number | null | undefined
  logPhoneNumberTypeEnabled: boolean
  googleMobileServicesAvailable: boolean | undefined
  huaweiMobileServicesAvailable: boolean | undefined
  pincodeUseExpandedBlocklist: boolean
  rewardPillText?: {
    [lang: string]: string
  }
  cashInButtonExpEnabled: boolean
  rampCashInButtonExpEnabled: boolean
  multiTokenShowHomeBalances: boolean
  multiTokenUseSendFlow: boolean
  multiTokenUseUpdatedFeed: boolean
  linkBankAccountEnabled: boolean
  sentryTracesSampleRate: number
  sentryNetworkErrors: string[]
  superchargeButtonType: SuperchargeButtonType
}

const initialState = {
  loading: false,
  loggedIn: false,
  numberVerified: false,
  analyticsEnabled: true,
  requirePinOnAppOpen: false,
  appState: AppState.Active,
  locked: false,
  lastTimeBackgrounded: 0,
  sessionId: '',
  minVersion: null,
  celoEducationUri: null,
  celoEuroEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.celoEuroEnabled,
  dappListApiUrl: null,
  inviteModalVisible: false,
  activeScreen: Screens.Main,
  hideVerification: REMOTE_CONFIG_VALUES_DEFAULTS.hideVerification,
  showRaiseDailyLimitTarget: undefined,
  walletConnectV1Enabled: REMOTE_CONFIG_VALUES_DEFAULTS.walletConnectV1Enabled,
  walletConnectV2Enabled: REMOTE_CONFIG_VALUES_DEFAULTS.walletConnectV2Enabled,
  rewardsPercent: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsPercent,
  rewardsStartDate: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsStartDate,
  rewardsMax: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsMax,
  rewardsMin: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsMin,
  rewardsABTestThreshold: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsABTestThreshold,
  ranVerificationMigrationAt: null,
  logPhoneNumberTypeEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.logPhoneNumberTypeEnabled,
  googleMobileServicesAvailable: undefined,
  huaweiMobileServicesAvailable: undefined,
  pincodeUseExpandedBlocklist: REMOTE_CONFIG_VALUES_DEFAULTS.pincodeUseExpandedBlocklist,
  rewardPillText: JSON.parse(REMOTE_CONFIG_VALUES_DEFAULTS.rewardPillText),
  cashInButtonExpEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.cashInButtonExpEnabled,
  rampCashInButtonExpEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.rampCashInButtonExpEnabled,
  multiTokenShowHomeBalances: REMOTE_CONFIG_VALUES_DEFAULTS.multiTokenShowHomeBalances,
  multiTokenUseSendFlow: REMOTE_CONFIG_VALUES_DEFAULTS.multiTokenUseSendFlow,
  multiTokenUseUpdatedFeed: REMOTE_CONFIG_VALUES_DEFAULTS.multiTokenUseUpdatedFeed,
  linkBankAccountEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.linkBankAccountEnabled,
  sentryTracesSampleRate: REMOTE_CONFIG_VALUES_DEFAULTS.sentryTracesSampleRate,
  sentryNetworkErrors: REMOTE_CONFIG_VALUES_DEFAULTS.sentryNetworkErrors.split(','),
  superchargeButtonType: REMOTE_CONFIG_VALUES_DEFAULTS.superchargeButtonType,
}

export const appReducer = (
  state: State | undefined = initialState,
  action: ActionTypes | RehydrateAction
): State => {
  switch (action.type) {
    case REHYDRATE: {
      // Ignore some persisted properties
      const rehydratePayload = getRehydratePayload(action, 'app')
      return {
        ...state,
        ...rehydratePayload,
        appState: initialState.appState,
        locked: rehydratePayload.requirePinOnAppOpen ?? initialState.locked,
        sessionId: '',
      }
    }
    case Actions.SET_APP_STATE:
      let { appState, lastTimeBackgrounded } = state
      switch (action.state) {
        case 'background':
          if (Platform.OS === 'android') {
            lastTimeBackgrounded = Date.now()
          }
          appState = AppState.Background
          break
        case 'inactive': // occurs only on iOS
          lastTimeBackgrounded = Date.now()
          appState = AppState.Inactive
          break
        case 'active':
          appState = AppState.Active
          break
      }
      return {
        ...state,
        appState,
        lastTimeBackgrounded,
      }
    case Actions.SET_LOGGED_IN:
      return {
        ...state,
        loggedIn: action.loggedIn,
      }
    case Actions.SET_NUMBER_VERIFIED:
      return {
        ...state,
        numberVerified: action.numberVerified,
      }
    case Actions.RESET_APP_OPENED_STATE:
      return {
        ...state,
        loggedIn: false,
        numberVerified: false,
      }
    case Actions.SET_ANALYTICS_ENABLED:
      return {
        ...state,
        analyticsEnabled: action.enabled,
      }
    case Actions.SET_LOCK_WITH_PIN_ENABLED:
      return {
        ...state,
        requirePinOnAppOpen: action.enabled,
      }
    case Actions.LOCK:
      return {
        ...state,
        locked: true,
      }
    case Actions.UNLOCK:
      return {
        ...state,
        locked: false,
      }
    case Actions.SET_SESSION_ID:
      return {
        ...state,
        sessionId: action.sessionId,
      }
    case Actions.MIN_APP_VERSION_DETERMINED:
      return {
        ...state,
        minVersion: action.minVersion,
      }
    case Actions.UPDATE_REMOTE_CONFIG_VALUES:
      return {
        ...state,
        hideVerification: action.configValues.hideVerification,
        showRaiseDailyLimitTarget: action.configValues.showRaiseDailyLimitTarget,
        celoEducationUri: action.configValues.celoEducationUri,
        celoEuroEnabled: action.configValues.celoEuroEnabled,
        dappListApiUrl: action.configValues.dappListApiUrl,
        walletConnectV1Enabled: action.configValues.walletConnectV1Enabled,
        walletConnectV2Enabled: action.configValues.walletConnectV2Enabled,
        rewardsPercent: action.configValues.rewardsPercent,
        rewardsStartDate: action.configValues.rewardsStartDate,
        rewardsMax: action.configValues.rewardsMax,
        rewardsMin: action.configValues.rewardsMin,
        rewardsABTestThreshold: action.configValues.rewardsABTestThreshold,
        logPhoneNumberTypeEnabled: action.configValues.logPhoneNumberTypeEnabled,
        pincodeUseExpandedBlocklist: action.configValues.pincodeUseExpandedBlocklist,
        rewardPillText: JSON.parse(action.configValues.rewardPillText),
        cashInButtonExpEnabled: action.configValues.cashInButtonExpEnabled,
        rampCashInButtonExpEnabled: action.configValues.rampCashInButtonExpEnabled,
        multiTokenShowHomeBalances: action.configValues.multiTokenShowHomeBalances,
        multiTokenUseSendFlow: action.configValues.multiTokenUseSendFlow,
        multiTokenUseUpdatedFeed: action.configValues.multiTokenUseUpdatedFeed,
        linkBankAccountEnabled: action.configValues.linkBankAccountEnabled,
        sentryTracesSampleRate: action.configValues.sentryTracesSampleRate,
        sentryNetworkErrors: action.configValues.sentryNetworkErrors,
        superchargeButtonType: action.configValues.superchargeButtonType,
      }
    case Actions.TOGGLE_INVITE_MODAL:
      return {
        ...state,
        inviteModalVisible: action.inviteModalVisible,
      }
    case Actions.ACTIVE_SCREEN_CHANGED:
      return {
        ...state,
        activeScreen: action.activeScreen,
      }
    case Actions.VERIFICATION_MIGRATION_RAN:
      return {
        ...state,
        ranVerificationMigrationAt: action.now,
        numberVerified: action.isVerified,
      }
    case Actions.ANDROID_MOBILE_SERVICES_AVAILABILITY_CHECKED:
      return {
        ...state,
        googleMobileServicesAvailable: action.googleIsAvailable,
        huaweiMobileServicesAvailable: action.huaweiIsAvailable,
      }
    default:
      return state
  }
}
