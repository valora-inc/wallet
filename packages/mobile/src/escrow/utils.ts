import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'
import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToE164NumberSelector } from 'src/identity/selectors'
import { ContactRecipient } from 'src/recipients/recipient'
import useSelector from 'src/redux/useSelector'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}

export const useEscrowPaymentRecipient = (payment: EscrowedPayment): ContactRecipient => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToE164Number = useSelector(identifierToE164NumberSelector)
  const recipientCache = useSelector((state) => state.recipients.phoneRecipientCache)
  const phoneNumber = identifierToE164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber] ?? {
    name: phoneNumber,
    e164PhoneNumber: phoneNumber,
    contactId: '',
  }

  return recipient
}
