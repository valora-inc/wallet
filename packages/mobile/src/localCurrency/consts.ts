// Supported local currency codes
// Please keep it sorted alphabetically
export enum LocalCurrencyCode {
  AUD = 'AUD',
  BRL = 'BRL',
  CAD = 'CAD',
  COP = 'COP',
  CVE = 'CVE',
  EUR = 'EUR',
  GBP = 'GBP',
  GHS = 'GHS',
  KES = 'KES',
  LRD = 'LRD',
  MXN = 'MXN',
  NGN = 'NGN',
  PHP = 'PHP',
  RUB = 'RUB',
  SLL = 'SLL',
  TRY = 'TRY',
  UAH = 'UAH',
  UGX = 'UGX',
  USD = 'USD',
}

export enum LocalCurrencySymbol {
  AUD = 'A$',
  BRL = 'R$',
  CAD = '$',
  COP = '$',
  CVE = '$',
  EUR = '€',
  GBP = '£',
  GHS = 'GH₵',
  KES = 'KSh',
  LRD = 'L$',
  MXN = '$',
  NGN = '₦',
  PHP = '₱',
  RUB = '₽',
  SLL = 'Le',
  TRY = '₺',
  UAH = '₴',
  UGX = 'USh',
  USD = '$',
}

export const LOCAL_CURRENCY_CODES = Object.values(LocalCurrencyCode)
