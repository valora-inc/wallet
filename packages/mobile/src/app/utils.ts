import { DappRequestOrigin } from 'src/analytics/types'
import { ActiveDapp } from 'src/app/reducers'

// Assume that if we have an active dapp, any WC or DappKit request comes from the in-app webview.
// Note that this may be incorrect if the user interacts with another dapp while the dapp webview is open.
// I thought about comparing the request url to the active dapp url, but that could fail too if the dapp url we have is different.
// For instance the dapp-list could contain a "example.com", while the request is "dapp.something.com".
export function getDappRequestOrigin(activeDapp: ActiveDapp | null) {
  if (activeDapp) {
    return DappRequestOrigin.InAppWebView
  }

  return DappRequestOrigin.External
}
