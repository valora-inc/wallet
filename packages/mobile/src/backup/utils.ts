import { MnemonicLanguages, normalizeMnemonic } from '@celo/utils/lib/account'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import { useDispatch, useSelector } from 'react-redux'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

export const DAYS_TO_BACKUP = 1
export const DAYS_TO_DELAY = 1 / 24 // 1 hour delay
export const MNEMONIC_SPLITTER = 'celo'

export const MNEMONIC_STORAGE_KEY = 'mnemonic'

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Account Key',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function formatBackupPhraseOnSubmit(phrase: string) {
  return normalizeMnemonic(phrase)
}

function isValidMnemonic(phrase: string, length: number) {
  return !!phrase && countMnemonicWords(formatBackupPhraseOnEdit(phrase)) === length
}

export function isValidBackupPhrase(phrase: string) {
  return isValidMnemonic(phrase, 24)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
