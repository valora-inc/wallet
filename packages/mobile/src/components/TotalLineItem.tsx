import BigNumber from 'bignumber.js'
import * as React from 'react'
import { Trans, useTranslation } from 'react-i18next'
import { StyleSheet, Text } from 'react-native'
import { MoneyAmount } from 'src/apollo/types'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import LineItemRow from 'src/components/LineItemRow'
import Touchable from 'src/components/Touchable'
import { useLocalCurrencyToShow } from 'src/localCurrency/hooks'
import { CurrencyInfo } from 'src/send/SendConfirmationLegacy'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import { Currency } from 'src/utils/currencies'

interface Props {
  title?: string
  amount: MoneyAmount
  hideSign?: boolean
  currencyInfo?: CurrencyInfo
  showExchangeRate?: boolean
  canEditCurrency?: boolean
  onEditCurrency?: () => void
}

const totalAmountKey = {
  [Currency.Dollar]: 'totalInDollars',
  [Currency.Euro]: 'totalInEuros',
  [Currency.Celo]: 'totalInCelo',
}

export default function TotalLineItem({
  title,
  amount,
  hideSign,
  currencyInfo,
  showExchangeRate = true,
  canEditCurrency = false,
  onEditCurrency,
}: Props) {
  const { localCurrencyExchangeRate: exchangeRate, amountCurrency } = useLocalCurrencyToShow(
    amount,
    currencyInfo
  )

  const { t } = useTranslation()

  return (
    <>
      <LineItemRow
        title={title || t('total')}
        textStyle={fontStyles.regular600}
        amount={
          <CurrencyDisplay
            amount={amount}
            hideSign={hideSign}
            currencyInfo={currencyInfo}
            testID="TotalLineItem/Total"
          />
        }
      />
      {showExchangeRate && exchangeRate && (
        <LineItemRow
          title={
            <Touchable disabled={!canEditCurrency} onPress={onEditCurrency}>
              <Text style={styles.exchangeRate}>
                <Trans i18nKey={totalAmountKey[amountCurrency]}>
                  <CurrencyDisplay
                    amount={{
                      value: new BigNumber(exchangeRate).pow(
                        amountCurrency === Currency.Celo ? 1 : -1
                      ),
                      currencyCode: Currency.Dollar, // The currency is actually the local amount
                    }}
                    showLocalAmount={false}
                    currencyInfo={currencyInfo}
                    testID="TotalLineItem/ExchangeRate"
                  />
                </Trans>
                {canEditCurrency && (
                  <>
                    {' '}
                    <Text style={styles.edit}>{t('edit')}</Text>
                  </>
                )}
              </Text>
            </Touchable>
          }
          amount={
            <CurrencyDisplay
              amount={amount}
              showLocalAmount={amountCurrency === Currency.Celo}
              hideSymbol={false}
              hideSign={hideSign}
              currencyInfo={currencyInfo}
              testID="TotalLineItem/Subtotal"
            />
          }
          style={styles.dollars}
          textStyle={styles.dollarsText}
        />
      )}
    </>
  )
}

const styles = StyleSheet.create({
  dollars: {
    marginVertical: 0,
  },
  dollarsText: {
    ...fontStyles.small,
    color: colors.gray4,
  },
  exchangeRate: {
    ...fontStyles.small,
    color: colors.gray4,
  },
  edit: {
    textDecorationLine: 'underline',
  },
})
