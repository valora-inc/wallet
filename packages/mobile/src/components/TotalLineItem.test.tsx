import { render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { MoneyAmount } from 'src/apollo/types'
import TotalLineItem from 'src/components/TotalLineItem'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { RootState } from 'src/redux/reducers'
import { CurrencyInfo } from 'src/send/SendConfirmationLegacy'
import { Currency } from 'src/utils/currencies'
import { createMockStore, RecursivePartial } from 'test/utils'

const defaultAmount: MoneyAmount = {
  value: '10',
  currencyCode: Currency.Dollar,
}
describe('TotalLineItem', () => {
  function renderComponent({
    storeOverrides = {},
    amount = defaultAmount,
    currencyInfo,
  }: {
    storeOverrides?: RecursivePartial<RootState>
    amount?: MoneyAmount
    currencyInfo?: CurrencyInfo
  }) {
    return render(
      <Provider
        store={createMockStore({
          localCurrency: {
            exchangeRates: { [Currency.Dollar]: null },
          },
          ...storeOverrides,
        })}
      >
        <TotalLineItem amount={amount} currencyInfo={currencyInfo} />
      </Provider>
    )
  }

  describe('when there is an exchange rate to show', () => {
    it('renders correctly', () => {
      const wrapper = renderComponent({
        currencyInfo: {
          localCurrencyCode: LocalCurrencyCode.PHP,
          localExchangeRate: '0.01',
        },
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe("when there isn't an exchange rate to show", () => {
    it("doesn't render the exchange rate", () => {
      const { queryByTestId } = renderComponent({})
      expect(queryByTestId('TotalLineItem/ExchangeRate')).toBeFalsy()
    })
  })
  describe("when there's an exchange rate to show in the currencyInfo", () => {
    it('renders the exchange rate', () => {
      const { queryByTestId } = renderComponent({
        currencyInfo: {
          localCurrencyCode: LocalCurrencyCode.PHP,
          localExchangeRate: '0.01',
        },
      })
      expect(queryByTestId('TotalLineItem/ExchangeRate/value')).toBeTruthy()
    })
  })
  describe("when there's an exchange rate to show in the amount", () => {
    it('renders the exchange rate', () => {
      const { queryByTestId } = renderComponent({
        amount: {
          ...defaultAmount,
          localAmount: {
            value: 10,
            currencyCode: LocalCurrencyCode.PHP,
            exchangeRate: 1.5,
          },
        },
      })
      expect(queryByTestId('TotalLineItem/ExchangeRate/value')).toBeTruthy()
    })
  })

  describe("when there's an exchange rate to show in the redux store", () => {
    it('renders the exchange rate', () => {
      const { queryByTestId } = renderComponent({
        storeOverrides: {
          localCurrency: {
            exchangeRates: { [Currency.Dollar]: '1.5' },
          },
        },
      })
      expect(queryByTestId('TotalLineItem/ExchangeRate/value')).toBeTruthy()
    })
  })
  describe('when the amount is in CELO', () => {
    it('displays the total in CELO and sub total in local currency', () => {
      const { getByText } = renderComponent({
        amount: {
          value: '10.12345',
          currencyCode: Currency.Celo,
          localAmount: {
            value: 15.185175,
            currencyCode: LocalCurrencyCode.EUR,
            exchangeRate: 1.5,
          },
        },
      })
      expect(getByText('10.123').props.testID).toEqual('TotalLineItem/Total/value')
      expect(getByText('€15.18').props.testID).toEqual('TotalLineItem/Subtotal/value')
      // TODO: This should use the local currency
      expect(getByText('$1.50').props.testID).toEqual('TotalLineItem/ExchangeRate/value')
      // expect(getByText('€1.50').props.testID).toEqual('TotalLineItem/ExchangeRate')
    })
  })
  describe('when the amount is in cUSD', () => {
    it('displays the total in local currency and sub total in cUSD', () => {
      const { getByText } = renderComponent({
        amount: {
          value: '10.12345',
          currencyCode: Currency.Dollar,
          localAmount: {
            value: 15.185175,
            currencyCode: LocalCurrencyCode.EUR,
            exchangeRate: 1.5,
          },
        },
      })
      expect(getByText('€15.18').props.testID).toEqual('TotalLineItem/Total/value')
      expect(getByText('$10.12').props.testID).toEqual('TotalLineItem/Subtotal/value')
      // TODO: This should use the local currency
      expect(getByText('$0.66').props.testID).toEqual('TotalLineItem/ExchangeRate/value')
      // expect(getByText('€0.66').props.testID).toEqual('TotalLineItem/ExchangeRate')
    })
  })
  describe('when the amount is in cEUR', () => {
    it('displays the total in local currency and sub total in cEUR', () => {
      const { getByText } = renderComponent({
        amount: {
          value: '10.12345',
          currencyCode: Currency.Euro,
          localAmount: {
            value: 15.185175,
            currencyCode: LocalCurrencyCode.PHP,
            exchangeRate: 1.5,
          },
        },
      })
      expect(getByText('₱15.18').props.testID).toEqual('TotalLineItem/Total/value')
      expect(getByText('€10.12').props.testID).toEqual('TotalLineItem/Subtotal/value')
      // FIXME: why is the rate showing with a $ sign?
      expect(getByText('$0.66').props.testID).toEqual('TotalLineItem/ExchangeRate/value')
    })
  })
})
