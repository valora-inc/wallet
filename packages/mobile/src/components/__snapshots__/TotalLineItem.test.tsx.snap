// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalLineItem when there is an exchange rate to show renders correctly 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginVertical": 5,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 5,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            },
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 16,
              "lineHeight": 22,
            },
          ]
        }
      >
        total
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#2E3338",
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 22,
          },
          Object {
            "color": "#2E3338",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 16,
            "lineHeight": 22,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": undefined,
            },
          ]
        }
        testID="TotalLineItem/Total/value"
      >
        
        $
        0.10
      </Text>
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginVertical": 5,
        },
        Object {
          "marginVertical": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 5,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            },
            Object {
              "color": "#9CA4A9",
              "fontFamily": "Inter-Regular",
              "fontSize": 14,
              "lineHeight": 18,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={false}
          nativeBackgroundAndroid={
            Object {
              "attribute": "selectableItemBackground",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Text
            style={
              Object {
                "color": "#9CA4A9",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": 18,
              }
            }
          >
            <Text
              style={
                Array [
                  undefined,
                  Object {
                    "color": undefined,
                  },
                ]
              }
              testID="TotalLineItem/ExchangeRate/value"
            >
              
              $
              100.00
            </Text>
          </Text>
        </View>
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#2E3338",
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 22,
          },
          Object {
            "color": "#9CA4A9",
            "fontFamily": "Inter-Regular",
            "fontSize": 14,
            "lineHeight": 18,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": undefined,
            },
          ]
        }
        testID="TotalLineItem/Subtotal/value"
      >
        
        $
        10.00
      </Text>
    </Text>
  </View>,
]
`;
